apps/cms/src/lib/cms/client/metadata-execution.ts

```diff
@@ -39,11 +39,94 @@ export const executeMetadataUpdate = async ({
   if (Object.keys(cmsProviderInstanceIdsToDelete).length) {
     await apiClient
       .mutation(DeleteMetadataDocument, {
-        id: productVariant.id,
+        id: itemId,
         keys: Object.entries(cmsProviderInstanceIdsToDelete).map(([cmsProviderInstanceId]) =>
           createCmsKeyForSaleorItem(cmsProviderInstanceId)
         ),
       })
       .toPromise();
   }
 };
+
+export const executeMetadataUpdate = async ({
+  context,
+  productVariant,
+  cmsProviderInstanceIdsToCreate,
+  cmsProviderInstanceIdsToDelete,
+}: {
+  context: WebhookContext;
+  productVariant: WebhookProductVariantFragment;
+  cmsProviderInstanceIdsToCreate: Record<string, string>;
+  cmsProviderInstanceIdsToDelete: Record<string, string>;
+}) => {
+  const { token, saleorApiUrl } = context.authData;
+  const apiClient = createClient(saleorApiUrl, async () => ({ token }));
+
+  await executeMetadataUpdateMutation({
+    apiClient,
+    itemId: productVariant.id,
+    cmsProviderInstanceIdsToCreate,
+    cmsProviderInstanceIdsToDelete,
+  });
+};
+
+type ItemMetadataRecord = {
+  id: string;
+  cmsProviderInstanceIds: MetadataRecord;
+};
+
+export const executeBatchMetadataUpdate = async ({
+  context,
+  variantCMSProviderInstanceIdsToCreate,
+  variantCMSProviderInstanceIdsToDelete,
+}: {
+  context: Pick<WebhookContext, "authData">;
+  variantCMSProviderInstanceIdsToCreate: ItemMetadataRecord[];
+  variantCMSProviderInstanceIdsToDelete: ItemMetadataRecord[];
+}) => {
+  const { token, saleorApiUrl } = context.authData;
+  const apiClient = createClient(saleorApiUrl, async () => ({ token }));
+
+  const variantCMSProviderInstanceIdsToCreateMap = variantCMSProviderInstanceIdsToCreate.reduce(
+    (acc, { id, cmsProviderInstanceIds }) => ({
+      ...acc,
+      [id]: {
+        ...(acc[id] || {}),
+        ...cmsProviderInstanceIds,
+      },
+    }),
+    {} as Record<string, MetadataRecord>
+  );
+  const variantCMSProviderInstanceIdsToDeleteMap = variantCMSProviderInstanceIdsToDelete.reduce(
+    (acc, { id, cmsProviderInstanceIds }) => ({
+      ...acc,
+      [id]: {
+        ...(acc[id] || {}),
+        ...cmsProviderInstanceIds,
+      },
+    }),
+    {} as Record<string, MetadataRecord>
+  );
+
+  await Promise.all(
+    Object.entries(variantCMSProviderInstanceIdsToCreateMap).map(
+      ([itemId, cmsProviderInstanceIdsToCreate]) =>
+        executeMetadataUpdateMutation({
+          apiClient,
+          itemId,
+          cmsProviderInstanceIdsToCreate,
+        })
+    )
+  );
+
+  await Promise.all(
+    Object.entries(variantCMSProviderInstanceIdsToDeleteMap).map(
+      ([itemId, cmsProviderInstanceIdsToDelete]) =>
+        executeMetadataUpdateMutation({
+          apiClient,
+          itemId,
+          cmsProviderInstanceIdsToDelete,
+        })
+    )
+  );
+};
```

Comment:
Are they supposed to run one after another or can they run in parallel as well?

Reply:
Can run parallel as well, changed.